### DISCUSSION TITLE
Discrepancy in LIBRIMIX Benchmark

### DISCUSSION THREAD

**Discussion by kaustubh14jr** [espnet/espnet#6097] [Q&A]
Dear @pengchengguo,

We were running ASR SOT conformer experiments on the Librimix dataset. We notice a difference in the number of words when compared to benchmark given on the github. We are getting exactly 3K words more while decoding on the test and dev set as compared to the benchmark on the github.

We are also unable to replicate the benchmark results while training the SOT conformer from scratch. Could there be a discrepancy in the number of words in the training set? We checked that the number of sentences tally with the Librimix paper. It looks difficult to tally the number of words or the content of the sentences.

Can someone please verify the training and decoding of SOT conformer script of LIBRIMIX dataset?

Thanks
KK

**Comment by sw005320** [comment]
Thanks for your discussion.
Can you elaborate on the explanations by listing the actual number of words in the GitHub and our generated data?

**Reply by kaustubh14jr** [reply]
Thanks for your quick response. We really appreciate it.
We would like to point you to this page:
https://github.com/espnet/espnet/tree/master/egs2/librimix/sot_asr1
There are three benchmarks on this page
[1] Conformer+wavlm: The number of words in the dev set and test set are 123853 and 111243 respectively. WER Dev 19.4 Test 17.1
[2] Only Conformer: The number of words in the dev set and test set are 123853 and 111243 respectively. WER Dev 24.7 WER Test 23.3
[3] SOT whisper_small: Here, the number of words are exactly 3K more ie 126853 for the dev set and 114243 for the test set. 
WER Dev 26.0 test 25.0

A point to note is that the WER for whisper_small is actually slightly worse than only conformer which we think is because of the extra 3K words. The benchmark[3] could be better than only conformer if the number of words were the same. Benchmark [3] was computed much later than [1] and [2].

In our experiments too, we are getting 3K more words as compared with benchmark [1] and [2]. We believe this is affecting both training and evaluation of our proposed method as well as the results we are trying to replicate. When we try to replicate benchmark [2], by training from scratch, we are getting the number of words exactly as benchmark [3] and WER Dev 35.6 Test 33.9 which is way off compared to benchmark [2]. 

We believe that the extra 3K words are because of the <sc> tag which is inserted in 3000 sentences of dev set and the test set. We cant figure out how it is affecting the training and decoding. We are working on the problem. If you could help us straighten out these benchmarks we would really appreciate it.

**Reply by kaustubh14jr** [reply]
I apologize for posting twice it is not allowing me to delete the comment.

**Comment by kaustubh14jr** [comment]
Thanks for your quick response. We really appreciate it.
We would like to point you to this page:
https://github.com/espnet/espnet/tree/master/egs2/librimix/sot_asr1
There are three benchmarks on this page
[1] Conformer+wavlm: The number of words in the dev set and test set are 123853 and 111243 respectively. WER Dev 19.4 Test 17.1
[2] Only Conformer: The number of words in the dev set and test set are 123853 and 111243 respectively. WER Dev 24.7 WER Test 23.3
[3] SOT whisper_small: Here, the number of words are exactly 3K more ie 126853 for the dev set and 114243 for the test set.
WER Dev 26.0 test 25.0

A point to note is that the WER for whisper_small is actually slightly worse than only conformer which we think is because of the extra 3K words. The benchmark[3] could be better than only conformer if the number of words were the same. Benchmark [3] was computed much later than [1] and [2].

In our experiments too, we are getting 3K more words as compared with benchmark [1] and [2]. We believe this is affecting both training and evaluation of our proposed method as well as the results we are trying to replicate. When we try to replicate benchmark [2], by training from scratch, we are getting the number of words exactly as benchmark [3] and WER Dev 35.6 Test 33.9 which is way off compared to benchmark [2].

We believe that the extra 3K words are because of the tag which is inserted in 3000 sentences of dev set and the test set. We cant figure out how it is affecting the training and decoding. We are working on the problem. If you could help us straighten out these benchmarks we would really appreciate it.

**Reply by sw005320** [reply]
Thanks for the detail, @kaustubh14jr!

@pengchengguo, could you check the details?

**Comment by pengchengguo** [comment]
Hi @kaustubh14jr,

Unfortunately, I am unavailable to the previous server now, so I need to re-download and prepare the data from scratch.
Once I have a conclusion, I will update you here.

**Reply by kaustubh14jr** [reply]
Hi @pengchengguo,
Thanks for your help. We will wait for your update.

**Comment by pengchengguo** [comment]
Hi @kaustubh14jr,

I think this difference may be caused by the `<sc>` token.
From the github commit https://github.com/espnet/espnet/commit/2ba9507c7e349155b719857ca555d961589cacb2, we can see a spacing is added after `<sc>`.
Before the commit, `<sc>` and the following word are combined as a one word, while after the commit, there are two different words during computing WERs.

**Reply by kaustubh14jr** [reply]
Hi @pengchengg and @sw005320,

@pengchengguo  Thanks for your help in identifying the issue.  

Will it be possible to retrain the models and update these two (Below) benchamarks and share the relevant yaml files.
1] Conformer+wavlm: The number of words in the dev set and test set are 123853 and 111243 respectively. WER Dev 19.4 Test 17.1
[2] Only Conformer: The number of words in the dev set and test set are 123853 and 111243 respectively. WER Dev 24.7 WER Test 23.3

When we try to recompute benchmark [2], with the corrected data preparation, we get the following result WER Dev 35.6 Test 33.9 which is lower than [2]. It is important for us because we are trying to propose a modification of the conformer and we have integrated our code into espnet so these benchmarks are very important to us. 

Thanks again for you help.

**Comment by pengchengguo** [comment]
Hi @kaustubh14jr,
We have uploaded the pre-trained models on Huggingface.
The fastest way to check the results is to add a "space" before and after the `<sc>` token and use the pre-trained model to conduct inference.

**Reply by kaustubh14jr** [reply]
Hi @pengchengguo ,

I copy paste the code from stage 5 local/data.sh where the commit was made. The code is to prepare the training, dev and eval transcripts. As you can see the commit also affects the training transcripts. If the training transcripts are also changing then, dont you think the pre-trained models on hugging face also will have to be retrained? Once you remove the concatenation \<sc> token with the next word there will be 64K extra words in the training transcripts. 64K is roughly the number of sentences in the train_460 which is the combination of  train_360 and train_100 of librispeech. With 64K extra words I think pretrained models for benchmark [1] and [2] will have to be retrained and updated. 

if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ]; then
    log "stage 5: LibriMix Transcripts Files Prepration."

    librispeech_data_dir="data/Librispeech"

    for dset in dev test train; do
        if [ "${dset}" = "train" ]; then
            src_ls_dset="train_clean_460"
        else
            src_ls_dset=${dset}"_clean"
        fi

        for i in $(seq 2); do
            awk -v idx=${i} '(FILENAME==ARGV[1]) {text[$1]=$0} (FILENAME==ARGV[2]) {split($1, lst, "_"); uttname=lst[idx]; print($1, text[uttname])}' \
                ${librispeech_data_dir}/${src_ls_dset}/text data/${dset}/wav.scp | \
                cut -d" " -f1,3- > data/${dset}/text_spk${i}
        done

        paste -d "" \
            <(<data/${dset}/text_spk1 awk '{$0=$0" <sc> "; print($0)}') \
            <(<data/${dset}/text_spk2 cut -d" " -f 2-) > data/${dset}/text

    done

fi

**Comment by pengchengguo** [comment]
Hi @kaustubh14jr,

1. The uploaded Conformer and Conformer + WavLM modes are trained prior to the commit https://github.com/espnet/espnet/commit/2ba9507c7e349155b719857ca555d961589cacb2. They used the text without a space between `<sc>` and the following word.
2. If checkout the codebase before commit https://github.com/espnet/espnet/commit/2ba9507c7e349155b719857ca555d961589cacb2, both training and inference with the huggingface model will reproduce the originally reported results.
3. After Chenda's commit, the training and evaluation text were modified to include an additional spacing after `<sc>`. As a result, the Whisper results can not be directly compared with previous benchmarks.
4. Many thanks if you could help to provide/or make a PR about Conformer results on the current commit or some Whisper results on the previous codebase.
5. I will also help update the new Conformer results and models later, but I’m currently unavailable.

**Reply by kaustubh14jr** [reply]
Hi @pengchengguo ,
Many thanks for all the clarifications. Because of the detailed discussion we can proceed with writing our paper. We can make a PR about the Conformer results on the current commit but you have to give us a few days. Thanks again for the discussion.
